storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,24 +3580,339 @@ dberr_t fsp_sys_tablespace_validate()
 }
 #endif /* UNIV_DEBUG */
 
-void fsp_system_tablespace_truncate()
+/** Store the inode information which basically stores
+the page and offset */
+struct inode_info : private std::unordered_set<uint64_t>
+{
+public:
+  __attribute__((warn_unused_result))
+  /** Register an inode */
+  bool insert_seg(const byte *inode)
+  {
+    uint32_t page= mach_read_from_4(inode + 4);
+    uint32_t offset= mach_read_from_2(inode + 8);
+    return page >= fil_system.sys_space->free_limit ||
+      offset < FIL_PAGE_DATA || offset >= srv_page_size - FIL_PAGE_DATA_END ||
+      emplace(uint64_t{page} << 16 | offset).second;
```

Comment:
The return value is not documented. According to this [description of std::unordered_set::emplace()](https://cplusplus.com/reference/unordered_set/unordered_set/emplace/), the `.second` component is `true` when the element to insert did not previously exist. So, the "successful" return value ought to be `true`. But the callers currently seem to expect `false`!

The conditions before `||` seem to be backwards. We would seem to want the following:
```c++
    return page < fil_system.sys_space->free_limit &&
      offset >= FIL_PAGE_DATA && offset < srv_page_size - FIL_PAGE_DATA_END &&
      emplace(uint64_t{page} << 16 | offset).second;
```